### PR TITLE
Repo Branch Rules page

### PR DESCRIPTION
**Pull request by sans-harness**
- Adds `repo-branch-rules` page to playground. 
- The svg for `toggle-button-inactive` is a placeholder icon for now. The correct icon will be added later, doesn't exist in the mocks for now. 

https://github.com/user-attachments/assets/23a5216c-bdcc-4b63-abb6-96384191f044

### CODE REVIEW

File: packages/canary/src/components/icon.tsx [hunk 1]
```diff
@@ -185,7 +187,9 @@ const IconNameMap = {
   'ssh-key': SshKey,
   'green-dot': GreenCircle,
   'green-tick': GreenTick,
-  'cancel-grey': CancelGrey
+  'cancel-grey': CancelGrey,
+  'toggle-active': ToggleActive,
```
Comment (abhinavrastogi-harness) [inline packages/canary/src/components/icon.tsx:191]: let's use the `<Switch>` component instead

File: packages/playground/src/pages/repo-settings-rules-page.tsx [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+  }
+  return (
+    <>
+      <form onSubmit={handleSubmit(onSubmit)}>
+        <FormFieldSet.Root>
+          <StackedList.Root className="border-none mb-8">
+            <StackedList.Item disableHover isHeader>
+              <StackedList.Field
+                title="Enable the rule"
+                description="Eget vestibulum eu lacinia ultrices id et morbi."
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/pages/repo-settings-rules-page.tsx:131]: let's file a design ticket on Rohan for providing the real text

File: packages/playground/src/pages/repo-settings-rules-page.tsx [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                  {...register('targetPatterns')}
+                  placeholder="Enter the target patterns"
+                  autoFocus
+                />
+                <Text size={2} as="p" color="tertiaryBackground" className="max-w-[100%]">
+                  Match branches using globstar patterns (e.g.”golden”, “feature-*”, “releases/**”)
+                </Text>{' '}
+              </div>
+              <div className="flex-[1]">
+                <Select value={includePatterns} onValueChange={value => handleSelectChange('includePatterns', value)}>
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/pages/repo-settings-rules-page.tsx:189]: this dropdown will only have `include/exclude` options. it should be a split button instead, similar to PR approve/reject button.

File: packages/playground/src/pages/repo-settings-rules-page.tsx [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  )
+})
+export type FormFields = z.infer<typeof formSchema>
+// interface RepoSettingsRulesPageProps {
+//   onFormSubmit?: (data: FormFields) => void
+//   onFormCancel?: () => void
+//   apiError?: string | null
+//   isLoading?: boolean
+//   isSuccess?: boolean
+// }
```
Comment (cjlee01) [inline packages/playground/src/pages/repo-settings-rules-page.tsx:48]: did we need this comment?

Comment (sans-harness) [reply packages/playground/src/pages/repo-settings-rules-page.tsx:48]: Yeah will be used when integrating to `gitness`

File: packages/playground/src/pages/repo-settings-rules-page.tsx [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+  },
+  {
+    id: 'status-checks',
+    label: 'Require status checks to pass',
+    description: 'Selected status checks must pass before a pull request can be merged'
+  },
+  {
+    id: 'merge-strategies',
+    label: 'Limit merge strategies',
+    description: 'Limit which merge strategies are available to merge a pull request'
```
Comment (cjlee01) [inline packages/playground/src/pages/repo-settings-rules-page.tsx:73]: this rule is more than a boolean

Comment (cjlee01) [reply packages/playground/src/pages/repo-settings-rules-page.tsx:73]: image

File: packages/playground/src/pages/repo-settings-rules-page.tsx [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+  },
+  {
+    id: 'comment-resolution',
+    label: 'Require comment resolution',
+    description: 'All comments on a pull request must be resolved before it can be merged'
+  },
+  {
+    id: 'status-checks',
+    label: 'Require status checks to pass',
+    description: 'Selected status checks must pass before a pull request can be merged'
```
Comment (cjlee01) [inline packages/playground/src/pages/repo-settings-rules-page.tsx:68]: image
this one would also be more than a boolean

**Review by cjlee01 — CHANGES_REQUESTED** [review]

File: packages/playground/src/pages/repo-settings-rules-page.tsx (hunk 6 above)
Comment (sans-harness) [reply packages/playground/src/pages/repo-settings-rules-page.tsx:68]: Oh interesting. 

The figma gave options for a boolean. I can include the changes but we would need design approvals for this as well. 

@abhinavrastogi-harness 


Screenshot 2024-10-17 at 4 12 56 PM

Comment (cjlee01) [reply packages/playground/src/pages/repo-settings-rules-page.tsx:68]: that wont work with the current api then and it doesnt even make sense if u think about merge strategies as a rule, you would need an input or some selected value cc. @sans-harness  @abhinavrastogi-harness

Comment (sans-harness) [reply packages/playground/src/pages/repo-settings-rules-page.tsx:68]: Oh yeah - I am implementing it, its a WIP, I just meant we needed to get design approvals. 

Current Progress~

https://github.com/user-attachments/assets/d1176c32-2eb0-4eb2-ade9-fce1c3c56f8f

**Comment by sans-harness** [comment]
Updated e2e vid:


Uploading e2e.mov…

**Comment by sans-harness** [comment]
e2e vid:


https://github.com/user-attachments/assets/750c3c97-b933-4dca-b873-847b09c176b8

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+  <FormFieldSet.ControlGroup>
+    <FormFieldSet.Label htmlFor="bypassValue">Bypass list</FormFieldSet.Label>
+    <Select
+      value={watch!('bypassValue')}
+      onValueChange={value => setValue!('bypassValue', value, { shouldValidate: true })}>
+      <SelectTrigger id="bypassValue">
+        <SelectValue placeholder="Select users" />
+      </SelectTrigger>
+      <SelectContent>
+        <SelectItem value="1">user 1</SelectItem>
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx:167]: why do we have hardcoded data in the reusable component? this data should be passed in via a prop

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    )}
+  </FormFieldSet.ControlGroup>
+)
+
+export const BranchSettingsRuleListField: React.FC<FieldProps> = ({ watch, setValue }) => {
+  const handleCheckboxChange = (ruleId: string, checked: boolean) => {
+    const updatedRules = watch!('rules').map(r => (r.id === ruleId ? { ...r, checked } : r))
+    setValue!('rules', updatedRules)
+  }
+  const handleSubmenuChange = (ruleId: string, submenuId: string, checked: boolean) => {
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx:203]: not a fan of manually maintaining form state in the components. few alternatives to consider:

1. use `useReducer` and move it out of the component
2. let the form store the raw data, and do this processing once when the form is being submitted

File: packages/playground/src/pages/repo-branch-settings-rules-page.tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  BranchSettingsRuleListField
+} from '../components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields'
+import { FormFieldSet } from '../index'
+import { branchRules } from '../components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-data'
+import {
+  RepoBranchSettingsFormFields,
+  repoBranchSettingsFormSchema
+} from '../components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-schema'
+
+// interface RepoSettingsRulesPageProps {
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/pages/repo-branch-settings-rules-page.tsx:22]: let's remove commented code if it's not needed

File: packages/playground/src/pages/repo-branch-settings-rules-page.tsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+export const RepoBranchSettingsRulesPage: React.FC<{ isLoading?: boolean }> = ({ isLoading = false }) => {
+  const {
+    register,
+    handleSubmit,
+    setValue,
+    watch,
+    reset,
+    formState: { errors, isValid }
+  } = useForm<RepoBranchSettingsFormFields>({
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/pages/repo-branch-settings-rules-page.tsx:38]: can we use `useZodForm` instead to simplify this setup?

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import { Rule, Action } from '../types'
+
+export const branchSettingsReducer = (state: Rule[], action: Action): Rule[] => {
+  switch (action.type) {
+    case 'TOGGLE_RULE':
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts:5]: can you make this into an enum?

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      return state.map(rule => (rule.id === action.ruleId ? { ...rule, checked: action.checked } : rule))
+    case 'TOGGLE_SUBMENU':
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts:7]: can you make this into an enum?

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+      return state.map(rule => {
+        if (rule.id === action.ruleId) {
+          const updatedSubmenu = action.checked
+            ? [...(rule.submenu || []), action.submenuId]
+            : (rule.submenu || []).filter(id => id !== action.submenuId)
+          return { ...rule, submenu: updatedSubmenu }
+        }
+        return rule
+      })
+    case 'SET_SELECT_OPTION':
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/reducers/repo-branch-settings-reducer.ts:17]: can you make this into an enum?

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+      <StackedList.Field
+        title="Enable the rule"
+        description="By enabling the toggle, the branch rule will be enforced."
+      />
+      <StackedList.Field
+        label
+        secondary
+        title={
+          <div className="flex gap-1.5 items-center justify-end cursor-pointer">
+            {' '}
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx:38]: remove this {' '}

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+        <div className="flex-[2.5]">
+          <Input
+            id="target-patterns"
+            {...register!('targetPatterns')}
+            placeholder="Enter the target patterns"
+            autoFocus
+          />
+          <Text size={2} as="p" color="tertiaryBackground" className="max-w-[100%] mt-2">
+            Match branches using globstar patterns (e.g.”golden”, “feature-*”, “releases/**”)
+          </Text>{' '}
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx:94]: remove this {' '}

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+        </div>
+        <Button
+          onClick={() => {}}
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/repo-branch-settings-rules-fields.tsx:97]: shud this onclick be empty funct?

File: packages/playground/src/components/repo-settings/repo-branch-settings-rules/types.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  id: string
+  checked: boolean
+  submenu: string[]
+  selectOptions: string
+}
+
+export type Action =
+  | { type: 'TOGGLE_RULE'; ruleId: string; checked: boolean }
+  | { type: 'TOGGLE_SUBMENU'; ruleId: string; submenuId: string; checked: boolean }
+  | { type: 'SET_SELECT_OPTION'; ruleId: string; selectedOptions: string }
```
Comment (cjlee01) [inline packages/playground/src/components/repo-settings/repo-branch-settings-rules/types.ts:14]: probably could put it into an enum if used here like above

File: packages/playground/src/pages/repo-branch-settings-rules-page.tsx [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+          <FormFieldSet.Root>
+            <FormFieldSet.ControlGroup>
+              <ButtonGroup.Root>
+                {!isSubmitted ? (
+                  <>
+                    <Button type="submit" size="sm" disabled={!isValid || isLoading}>
+                      {!isLoading ? 'Create rule' : 'Creating rule...'}
+                    </Button>
+                    <Button type="button" variant="outline" size="sm" onClick={() => {}}>
```
Comment (cjlee01) [inline packages/playground/src/pages/repo-branch-settings-rules-page.tsx:90]: empty onclick?

**Review by cjlee01 — APPROVED** [review]

**Pull request merged**